### ISSUE TITLE
Type inference for implicit val fails with a strange error

### ISSUE THREAD

**Issue by KuceraMartin** [lampepfl/dotty#19373]
## Compiler version
3.3.0, 3.4.0-RC1-bin-20240103-94e960d-NIGHTLY

## Minimized code
```Scala
trait Box

given Conversion[String, Box] = ???

def unbox(b: Box): String = ???

def foo: Unit =
  val s = unbox("foo")
  implicit val is = s
```

## Output
```
[error] Recursive value s needs type
[error]   implicit val is = s
[error]                     ^
Warning: there was 1 feature warning; re-run with -feature for details
Error compiling project (Scala 3.3.0, JVM)
Compilation failed
```

Tree after typer:
```scala
package <empty> {
  trait Box() extends Object {}
  final lazy module val recursive-value-bug$package: recursive-value-bug$package
     = new recursive-value-bug$package()
  final module class recursive-value-bug$package() extends Object() {
    this: recursive-value-bug$package.type =>
    final lazy given val given_Conversion_String_Box: Conversion[String, Box] =
      ???
    def unbox(b: Box): String = ???
    def foo: Unit =
      {
        val s: String = unbox(given_Conversion_String_Box.apply("foo"))
        implicit val is: <error Recursive value s needs type> = s
        ()
      }
  }
}
```
Note that somehow the implicit conversion from `String` to `Box` is important; replacing the `unbox` call with `identity` fixes the problem. 

Strangely, giving an explicit type to `s` also fixes the problem, even though the type of `s` was already inferred correctly before:
```scala
def foo: Unit =
  val s: String = unbox("foo")
  implicit val is = s // ok
```

Giving an explicit type to `is` also fixes the problem:
```scala
def foo: Unit =
  val s = unbox("foo")
  implicit val is: String = s // ok
```

But when typing `s` inside the val assignment, we get the same error as before:
```scala
def foo: Unit =
  val s = unbox("foo")
  implicit val is = (s: String) // error: Recursive value s needs type
```

## Expectation
It should probably compile, or otherwise give a more meaningful error.

**Comment by odersky** [comment]
Well, here you could just give `s` a type and proceed, so the error message is slightly wrong, but still helpful.

What happens:

 - The compiler tries to type `val s = unbox("foo")`. This fails since `"foo"` is not a `Box`. 
 - The compiler looks for implicit conversions. The next implicit in scope is `is`, but to figure out whether that's a conversion or some other implicit, the compiler has to determine its type.
 - Since `is` does not have an explicit type, we need to typecheck its right hand side, `s`.

Which gives a cycle.

Givens don't have this problem since the type is always explicit. And, please, let's all stop using old-style implicits in Scala 3.

**Comment by odersky** [comment]
I put this in backlog since it relies on old-style implicits which will go away anyway.

**Comment by KuceraMartin** [comment]
Thanks for the explanation @odersky, now I understand why the problem occurs.

As for old-style implicits in Scala 3. Is there a way to avoid them if I want to cross-compile to Scala 2?

**Comment by odersky** [comment]
No, then you are stuck. But maybe we should start doing Scala 3 only?

**Comment by KuceraMartin** [comment]
That would be ideal. Unfortunately we still have a lot of Scala 2 projects at Ematiq, and migrating them is not a priority (not my decision).

This is probably not the right place but I can share more about our issues with migration in another channel if you’re interested.
